### ISSUE TITLE
[FEATURE] @EqualsAndHashCode for records

### ISSUE THREAD

**Issue by pawelkedra** [projectlombok/lombok#3246]
**Describe the feature**
Yes, I know - `equals()` and `hashcode()` are automatically generated for records, but there's no way to customise them. Sometimes you don't want to compare all object properties, e.g. in case of DB entities it's usually enough to implement equals/hashcode using only the ID. In case of classes, it's very easy to customise it with Lombok:
```java
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class MyEntity {
    @Id
    @EqualsAndHashCode.Include
    private UUID id;
    private String name;     
}
```
Unfortunately, I can't do the same with records - they are not supported by Lombok's `@EqualsAndHashCode` annotation. This piece of code fails to compile with `java: @EqualsAndHashCode is only supported on a class.` error:
```java
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public record MyEntity (
    @Id 
    @EqualsAndHashCode.Include
    UUID id,
    String name
) {}
```
I think it would be great to have this annotation also for records, and throw the compilation error only if `EqualsAndHashCode.Include/Exclude` is not used in the code.

**Describe the target audience**
Anyone who migrates POJO classes to Java records, but has specific requirements regarding `equals()` and `hashcode()` methods.

**Comment by swaranga** [comment]
I understand the intent of the request to have Lombok generate the equals and hashcode via annotations but it is not accurate to say there is no way to customize the equals and hashcode method for record classes. You can definitely overridde them and implement your own logic

```java
public record MyEntity(UUID id, String name) {
    @Override
    public boolean equals(Object o) {
        return o instanceof MyEntity other &&
                   Objects.equals(this.id, other.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(this.id);
    }
}
```

**Comment by pawelkedra** [comment]
Well, yes, but that's true also for classes, which are supported by @EqualsAndHashCode.

**Comment by swaranga** [comment]
Agreed. I was only responding to your comment about record classes: 
> `but there's no way to customize them`. 

Not disputing the utility of the feature request itself.

**Comment by pawelkedra** [comment]
Updated the description: `but there's no easy way to customize them.`

**Comment by datze** [comment]
There is another reason for a `@EqualsAndHashCode` annotation on a record: When the record contains an array, I would expect the Lombok-generated equals() and hashcode() methods to use the arrays values for comparison (with `Arrays.equals()` instead of `o1.equals(o2)`).

**Comment by TheConen** [comment]
Exactly that. After an update from Sonar 8 LTS to Sonar 9 LTS I had some new findings about https://rules.sonarsource.com/java/RSPEC-6218 and was expecting to be able to use @EqualsAndHashCode for that - and was very surprised when that was not possible.

**Comment by luneo7** [comment]
One other use for that is https://github.com/oracle/graal/issues/4348 ... Where in native image right now at least records own implementation lacks performance since they are doing reflection... I know that they will be improving it in a new GraalVM release, but a "handwritten" method (`@ToString` as well) will definitely improve performance.
It would be good to support records with `@EqualsAndHashCode` and `@ToString` as well...

**Comment by ukv001** [comment]
Another place where it would be good to have the support of the annotations is the exclude. e.g. when writing a test, using mockito to verify the content of an instance that is called on a specific method, where a timestamp is included, then the equals.exclude would be very useful.

**Comment by cerebro84** [comment]
> Another place where it would be good to have the support of the annotations is the exclude. e.g. when writing a test, using mockito to verify the content of an instance that is called on a specific method, where a timestamp is included, then the equals.exclude would be very useful.

@ukv001 For that you don't normally want to override equals, you can use assertJ and, to make sure you really compare the fields and you don't rely on the implemented equals, use `.usingRecursiveComparison()` combined with `.ignoringFields("nameOfTheField")`.

**Comment by nlisker** [comment]
I'm confused about the lack of support on records because the annotations themselves define that their target is a type:
```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.SOURCE)
public @interface EqualsAndHashCode {
```
and `ElementType.TYPE`'s docs say:
"Class, interface (including annotation interface), enum, or record declaration"
So they *should* work on records.

**Comment by rzwitserloot** [comment]
@nlisker Yes, you are confused. `@Target` doesn't identify what's allowed. It identifies what isn't allowed. It's not the same thing.

Javac _itself_ can restrict annotations; putting them in a place that is disallowed _as per `@Target` will cause `javac` itself to generate an error, and the annotation processor isn't even called.

However, annotation processors can go a lot further. They can generate errors on their own. Trivially, `@Override` can go on methods and yet, if you slap that on a method that doesn't override anything, a compiler error ensues. And yet, `@Target(ElementType.OVERRIDDEN_METHOD)` is not on on `@interface Override`. In fact, `OVERRIDDEN_METHOD` doesn't exist at all.

Lombok is the same. If there was a `ElementType.EVERY_TYPE_EXCEPT_RECORDS`, we'd have stuck that on `@interface EqualsAndHashCode` but it doesn't exist.

The point is, the `@Target` annotation is a negative. It defines what is not allowed. That does not imply that what remains _is_ allowed. That's not a lombok specific thing, that's just how annotations work.

**Comment by rzwitserloot** [comment]
I think when we added `record` support we were under the mistaken belief that `equals` and `hashCode` were more or less hardcoded, more or less in the same way the constructor is.

Clearly that's not true.

We're probably going to continue to disallow `@Data` and `@Value`, but an explicit `@EqualsAndHashCode` seems reasonable.

**Comment by nlisker** [comment]
> We're probably going to continue to disallow `@Data` and `@Value`, but an explicit `@EqualsAndHashCode` seems reasonable.

I believe that only `@EqualsAndHashCode`, `@ToString`, and `@Getter` are viable (from those that aren't already supported).
